### PR TITLE
DBZ-7258 Enable Retry on the Debezium Storage JDBC

### PR DESCRIPTION
**Pull request by ilyasahsan123**
https://issues.redhat.com/browse/DBZ-7258

### CODE REVIEW

**Review by sherpa003 — CHANGES_REQUESTED** [review]

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/offset/JdbcOffsetBackingStore.java [hunk 1]
```diff
@@ -179,7 +179,7 @@ public synchronized void stop() {
         stopExecutor();
         try {
             if (conn != null) {
-                conn.close();
+                conn.connection().close();
```
[the remark this answers is not in the document]
Comment (ilyasahsan123) [reply debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/offset/JdbcOffsetBackingStore.java:182]: I'll consider it. Thanks for your recommendation.

File: debezium-storage/debezium-storage-jdbc/src/test/java/io/debezium/storage/jdbc/JdbcConnectionIT.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    @After
+    public void afterEach() {
+        stopConnector();
+    }
+
+    @Test
+    public void shouldEstablishConnectionWithRetry() {
+        String jdbcUrl = String.format("jdbc:mysql://%s:%s/%s", container.getHost(), container.getMappedPort(PORT), DBNAME);
+        try (JdbcConnection conn = new JdbcConnection(jdbcUrl, PRIVILEGED_USER, PRIVILEGED_PASSWORD, WAIT_RETRY_DELAY, MAX_RETRY_COUNT)) {
+            conn.close();
```
[the remark this answers is not in the document]
Comment (ilyasahsan123) [reply debezium-storage/debezium-storage-jdbc/src/test/java/io/debezium/storage/jdbc/JdbcConnectionIT.java:71]: Got it, I'll give your suggestion a shot and manually test the scenario.

[the remark this answers is not in the document]
Comment (ilyasahsan123) [reply debezium-storage/debezium-storage-jdbc/src/test/java/io/debezium/storage/jdbc/JdbcConnectionIT.java:71]: hi @jpechane, do you have any suggestion for this?

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        this.waitRetryDelay = waitRetryDelay;
+        this.maxRetryCount = maxRetryCount;
+        this.conn = establishConnectionWithRetry();
+    }
+
+    public synchronized boolean isConnected() throws SQLException {
+        if (conn == null) {
+            return false;
+        }
+        return !conn.isClosed();
```
[the remark this answers is not in the document]
Comment (ilyasahsan123) [reply debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java:46]: Got it. I'll try your recommendation. Thanks for allowing me to continue this pull request.

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcCommonConfig.java [hunk 4]
```diff
@@ -39,9 +39,21 @@ public class JdbcCommonConfig {
             .withDescription("Password of the database which will be used to access the database storage")
             .withValidation(Field::isRequired);
 
+    private static final long DEFAULT_WAIT_RETRY_DELAY = 1000L;
+    private static final Field PROP_WAIT_RETRY_DELAY = Field.create(CONFIGURATION_FIELD_PREFIX_STRING + "wait.retry.delay.ms")
+            .withDescription("Delay of retry on wait for connection failure")
+            .withDefault(DEFAULT_WAIT_RETRY_DELAY);
+
+    private static final Integer DEFAULT_MAX_RETRIES = 10;
```
Comment (jpechane) [inline debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcCommonConfig.java:47]:
```suggestion
    private static final int DEFAULT_MAX_RETRIES = 10;
```

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcCommonConfig.java [hunk 5]
```diff
@@ -39,9 +39,21 @@ public class JdbcCommonConfig {
...
+    private static final Field PROP_MAX_RETRIES = Field.create(CONFIGURATION_FIELD_PREFIX_STRING + "retry.max.attempts")
+            .withDescription("Maximum number of retry attempts before giving up.")
+            .withDefault(DEFAULT_MAX_RETRIES);
+
     private String jdbcUrl;
     private String user;
     private String password;
+    private long waitRetryDelay;
```
Comment (jpechane) [inline debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcCommonConfig.java:55]:
```suggestion
    private Duration waitRetryDelay;
```

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcCommonConfig.java [hunk 6]
```diff
@@ -74,4 +88,12 @@ public String getUser() {
     public String getPassword() {
         return password;
     }
+
+    public long getWaitRetryDelay() {
```
Comment (jpechane) [inline debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcCommonConfig.java:92]:
```suggestion
    public Duration getWaitRetryDelay() {
```

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcCommonConfig.java [hunk 7]
```diff
@@ -74,4 +88,12 @@ public String getUser() {
...
+        return waitRetryDelay;
+    }
+
+    public Integer getMaxRetryCount() {
```
Comment (jpechane) [inline debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcCommonConfig.java:96]:
```suggestion
    public int getMaxRetryCount() {
```

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java:46]: I'd recommend that there will be a template method that will accept a functional interface with JDBC code.
The template method will start TX, call the passed code and commit/rollback at the end.
If an exception is thrown the template method will recreate the connection and retry the execution.

**Review by jpechane — CHANGES_REQUESTED** [review]

**Comment by ilyasahsan123** [comment]
hi @jpechane @sherpa003, 
I've incorporated your suggestions. could you please take another look? 
Thanks

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        }
+        catch (SQLRecoverableException e) {
+            handleRecoverableException(e);
+        }
+        catch (SQLException e) {
+            LOGGER.error("Exception while executing SQL", e);
+            throw new ConnectException("Failed to execute SQL", e);
+        }
+        finally {
+            if (rollback) {
```
Comment (sherpa003) [inline debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java:56]: this is certainly wrong. Rollback should be done only within a catch Exception block (as you can see in my example: https://github.com/sherpa003/debezium/commit/6719ed1582dfc14fb58b6708a556f90d02ea638b)

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    private synchronized void rollback() {
+        try {
+            conn.rollback();
+        }
+        catch (SQLException e) {
+            // ignore rollback exception
+        }
+    }
+
+    private synchronized boolean isConnected() {
```
Comment (sherpa003) [inline debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java:91]: no need to have all private method synchronized when they are called from within another synchronized method

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        this.user = user;
+        this.password = password;
+        this.waitRetryDelay = waitRetryDelay;
+        this.maxRetryCount = maxRetryCount;
+        create();
+    }
+
+    public synchronized void execute(JdbcConnectionHandler handler, boolean rollback) {
+        try {
+            handler.apply(conn);
```
Comment (sherpa003) [inline debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java:46]: this is never executed again after the reconnect.

**Review by sherpa003 — CHANGES_REQUESTED** [review]
it seems to me that this will not work at all.

**Comment by sherpa003** [comment]
@ilyasahsan123 , could you please let me know if you are continuing on this or if I should try to create the PR by myself based on the code I shared?

**Comment by ilyasahsan123** [comment]
I apologize for delayed update on this PR. Please take over this and I will close this PR, @sherpa003 . 🙏

**Pull request closed without merging**